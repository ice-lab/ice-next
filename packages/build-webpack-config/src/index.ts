import * as path from 'path';
import type { Configuration } from 'webpack';
import type { Configuration as DevServerConfiguration } from 'webpack-dev-server';
import type { Config } from '@ice/types';
import swcPlugin from './swcPlugin.js';

const watchIgnoredRegexp = process.env.RUNTIME_DEBUG ? /node_modules/ : /node_modules|[/\\]\.ice[/\\]|[/\\]\.rax[/\\]/;

interface GetWebpackConfigOptions {
  rootDir: string;
  config: Config;
}
type GetWebpackConfig = (options: GetWebpackConfigOptions) => Configuration & { devServer?: DevServerConfiguration };

export const getWebpackConfig: GetWebpackConfig = ({ rootDir, config }) => {
  const {
    mode,
    externals = {},
    publicPath = '/',
    devPublicPath = '/',
    outputDir = path.join(rootDir, 'build'),
    loaders = [],
    alias = {},
    sourceMap,
    middlewares,
  } = config;

  return {
    mode,
    entry: path.join(rootDir, 'src/app'),
    externals,
    output: {
      publicPath,
      path: outputDir,
    },
    module: {
      rules: [
        ...loaders,
      ],
    },
    resolve: {
      alias: {
        ice: path.join(rootDir, '.ice', 'index.ts'),
        ...alias,
      },
      extensions: ['.ts', '.tsx', '.jsx', '...'],
    },
    plugins: [
      swcPlugin({ rootDir, sourceMap }),
    ],
    devtool: 'source-map',
    devServer: {
      allowedHosts: 'all',
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': '*',
        'Access-Control-Allow-Headers': '*',
      },
      hot: true,
      compress: true,
      webSocketServer: 'ws',
      devMiddleware: {
        publicPath: devPublicPath,
      },
      static: {
        watch: {
          ignored: watchIgnoredRegexp,
        },
      },
      client: {
        overlay: false,
        logging: 'info',
      },
      setupMiddlewares: middlewares,
    },
  };
};
