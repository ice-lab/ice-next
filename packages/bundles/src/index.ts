// reexport for dependencies
import postcss from 'postcss';
import less from 'less';
import sass from 'sass';

export {
  postcss,
  less,
  sass,
};

export type { ProcessOptions } from 'postcss';
