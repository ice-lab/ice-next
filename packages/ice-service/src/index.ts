import { Context } from 'build-scripts';
import type webpack from 'webpack';
import type { CommandArgs, CommandName, IPluginAPI, IGetBuiltInPlugins } from 'build-scripts';
import Generator from './service/runtimeGenerator';
import preCompile from './service/preCompile';
import start from './commands/start';
import build from './commands/build';
import type { ExportData } from './service/runtimeGenerator';
import type { Config } from '@builder/webpack-config';

type AddExport = (exportData: ExportData) => void;

interface ExtendsPluginAPI {
  context: {
    // TODO define routeManifest
    routeManifest: any;
    webpack?: typeof webpack;
  };
  generator: {
    addExport: AddExport;
    addExportTypes: AddExport;
    addConfigTypes: AddExport;
    addRenderFile: Generator['addRenderFile'];
    addRenderTemplate: Generator['addTemplateFiles'];
  };
}

export interface FrameworkPlugin<T = undefined> {
  (api: IPluginAPI<Config, ExtendsPluginAPI>, options?: T): Promise<void> | void;
}

interface CreateServiceOptions {
  rootDir: string;
  command: CommandName;
  commandArgs: CommandArgs;
  getBuiltInPlugins: IGetBuiltInPlugins;
}

async function createService({ rootDir, command, commandArgs, getBuiltInPlugins }: CreateServiceOptions) {
  // TODO pre compile
  preCompile();
  const routeManifest = {};
  const generator = new Generator({
    rootDir,
    targetDir: './.ice',
     // TODO get default Data
    defaultData: {},
  });
  const generatorAPI = {
    addExport: (exportData: ExportData) => {
      generator.addExport('framework', exportData);
    },
    addExportTypes: (exportData: ExportData) => {
      generator.addExport('frameworkTypes', exportData);
    },
    addConfigTypes: (exportData: ExportData) => {
      generator.addExport('configTypes', exportData);
    },
    addRenderFile: generator.addRenderFile,
    addRenderTemplate: generator.addTemplateFiles,
  };
  const ctx = new Context<any, ExtendsPluginAPI>({
    rootDir,
    command,
    commandArgs,
    extendsPluginAPI: {
      generator: generatorAPI,
      context: {
        routeManifest,
      },
    },
    getBuiltInPlugins,
  });
  await ctx.resolveConfig();
  generator.setPlugins(ctx.getAllPlugin());
  await ctx.setup();
  // render template before webpack compile
  generator.render();

  return {
    run: async () => {
      if (command === 'start') {
        return await start(ctx);
      } else if (command === 'build') {
        return await build(ctx);
      }
    },
  };
}

export default createService;
