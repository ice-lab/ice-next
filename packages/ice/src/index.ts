import path from 'path';
import { fileURLToPath } from 'url';
import { Context } from 'build-scripts';
import consola from 'consola';
import type { CommandArgs, CommandName, IGetBuiltInPlugins } from 'build-scripts';
import Generator from './service/runtimeGenerator.js';
import preCompile from './service/preCompile.js';
import createWatch from './service/watchSource.js';
import start from './commands/start.js';
import build from './commands/build.js';
import type { ExportData } from '@ice/types/esm/generator.js';
import type { ExtendsPluginAPI } from '@ice/types/esm/plugin.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

interface CreateServiceOptions {
  rootDir: string;
  command: CommandName;
  commandArgs: CommandArgs;
  getBuiltInPlugins: IGetBuiltInPlugins;
}

async function createService({ rootDir, command, commandArgs, getBuiltInPlugins }: CreateServiceOptions) {
  // TODO pre compile
  preCompile();
  const routeManifest = {};
  const generator = new Generator({
    rootDir,
    targetDir: './.ice',
     // TODO get default Data
    defaultData: {},
  });
  // add default template of ice
  const templatePath = path.join(__dirname, '../template/');
  generator.addTemplateFiles(templatePath);
  const generatorAPI = {
    addExport: (exportData: ExportData) => {
      generator.addExport('framework', exportData);
    },
    addExportTypes: (exportData: ExportData) => {
      generator.addExport('frameworkTypes', exportData);
    },
    addConfigTypes: (exportData: ExportData) => {
      generator.addExport('configTypes', exportData);
    },
    addRenderFile: generator.addRenderFile,
    addRenderTemplate: generator.addTemplateFiles,
  };
  const { addWatchEvent, removeWatchEvent } = createWatch(path.join(rootDir, 'src'));
  const ctx = new Context<any, ExtendsPluginAPI>({
    rootDir,
    command,
    commandArgs,
    extendsPluginAPI: {
      generator: generatorAPI,
      watch: {
        addEvent: addWatchEvent,
        removeEvent: removeWatchEvent,
      },
      context: {
        routeManifest,
      },
    },
    getBuiltInPlugins,
  });
  await ctx.resolveConfig();
  generator.setPlugins(ctx.getAllPlugin());
  await ctx.setup();
  // render template before webpack compile
  const renderStart = new Date().getTime();
  generator.render();
  consola.debug('template render cost:', new Date().getTime() - renderStart);

  return {
    run: async () => {
      if (command === 'start') {
        return await start(ctx);
      } else if (command === 'build') {
        return await build(ctx);
      }
    },
  };
}

export default createService;
