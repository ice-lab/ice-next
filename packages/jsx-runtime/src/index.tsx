import { jsx as _jsx, jsxs as _jsxs, Fragment } from 'react/jsx-runtime';
import { convertUnit } from 'style-unit';

const STYLE = 'style';

/**
 * https://github.com/reactjs/rfcs/pull/107
 * @param {*} type
 * @param {object} props
 * @param {string} key
 */
export function jsx(type: any, props: object, maybeKey: string) {
  return _jsx(type, hijackElementProps(props), maybeKey);
}

// Same as jsx method, special case jsxs internally to take advantage of static children.
// // for now we can ship identical prod functions.
export function jsxs(type: any, props: object, maybeKey: string) {
  return _jsxs(type, hijackElementProps(props), maybeKey);
}

function isObject(obj: any): obj is object {
  return typeof obj === 'object';
}

// Support rpx unit.
function hijackElementProps(props: { style?: object } | object): object {
  if (props && STYLE in props) {
    const { style } = props;
    if (isObject(style)) {
      const result = Object.assign({}, props);
      const convertedStyle = {};
      for (const prop in style) {
        // @ts-ignore
        convertedStyle[prop] = convertUnit(style[prop]);
      }
      result['style'] = convertedStyle;
      return result;
    }
  }
  return props;
}

export {
  Fragment,
};
