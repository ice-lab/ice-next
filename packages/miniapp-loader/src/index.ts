import type webpack from 'webpack';

import pageLoader from './page.js';

export default function (this: webpack.LoaderContext<any>, source: string) {
  pageLoader.call(this, source);
}
