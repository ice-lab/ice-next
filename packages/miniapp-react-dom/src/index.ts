/* eslint-disable @typescript-eslint/no-unused-vars */
import type { Element } from '@ice/miniapp-runtime';
import { ensure, isFunction } from '@ice/shared';
import type { ReactNode } from 'react';

import { IceMiniappReconciler } from './reconciler.js';
import { ContainerMap, createRoot, render } from './render.js';

const { batchedUpdates } = IceMiniappReconciler;

function unmountComponentAtNode(dom: Element) {
  ensure(dom && [1, 8, 9, 11].includes(dom.nodeType), 'unmountComponentAtNode(...): Target container is not a DOM element.');

  const root = ContainerMap.get(dom);

  if (!root) return false;

  batchedUpdates(() => {
    root.unmount(() => {
      ContainerMap.delete(dom);
    });
  }, null);

  return true;
}

function findDOMNode(comp?: Element | ReactNode) {
  if (comp == null) {
    return null;
  }

  const { nodeType } = comp as Element;
  if (nodeType === 1 || nodeType === 3) {
    return comp;
  }

  return IceMiniappReconciler.findHostInstance(comp as Record<string, any>);
}

const portalType = isFunction(Symbol) && Symbol.for
  ? Symbol.for('react.portal')
  : 0xeaca;

function createPortal(
  children: ReactNode,
  containerInfo: Element,
  key?: string,
) {
  return {
    $$typeof: portalType,
    key: key == null ? null : String(key),
    children,
    containerInfo,
    implementation: null,
  };
}

const Internals = {
  usingClientEntryPoint: false,
};

export {
  // Hack for import ReactDOM from 'react-dom/client'
  Internals as __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED,
  createPortal,
  createRoot,
  findDOMNode,
  render,
  unmountComponentAtNode,
  /* eslint-disable-next-line camelcase */
  batchedUpdates as unstable_batchedUpdates,
};

export default {
  render,
  createRoot,
  unstable_batchedUpdates: batchedUpdates,
  unmountComponentAtNode,
  findDOMNode,
  createPortal,
};
