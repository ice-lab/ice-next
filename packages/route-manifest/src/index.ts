// based on https://github.com/remix-run/remix/blob/main/packages/remix-dev/config/routesConvention.ts
import fs from 'fs';
import path from 'path';
import minimatch from 'minimatch';
import { createRouteId, defineRoutes } from './routes.js';
import type { RouteManifest, DefineRouteFunction, NestedRouteManifest } from './routes.js';

const routeModuleExts = ['.js', '.jsx', '.ts', '.tsx', '.md', '.mdx'];

export function isRouteModuleFile(filename: string): boolean {
  return routeModuleExts.includes(path.extname(filename));
}

export function generateRouteManifest(rootDir: string) {
  const srcDir = path.join(rootDir, 'src');
  const routeManifest: RouteManifest = {};
  // 1. find global layout
  const globalLayoutFile = findGlobalLayout(srcDir, 'layout');
  if (globalLayoutFile) {
    routeManifest['layout'] = { path: '', id: 'layout', file: globalLayoutFile };
  }

  // 2. find routes in `src/pages` directory
  if (fs.existsSync(path.resolve(srcDir, 'pages'))) {
    const conventionalRoutes = defineConventionalRoutes(
      rootDir,
      [], // TODO: add ignoredFilePatterns defined in ice.config.js
    );

    for (const key of Object.keys(conventionalRoutes)) {
      const route = conventionalRoutes[key];
      routeManifest[route.id] = {
        ...route,
        parentId: route.parentId || (globalLayoutFile && 'layout') || undefined,
      };
    }
  }

  return routeManifest;
}

export function formatNestedRouteManifest(routeManifest: RouteManifest, parentId?: string): NestedRouteManifest[] {
  return Object.keys(routeManifest)
    .filter(key => routeManifest[key].parentId === parentId)
    .map(key => {
      const route = {
        ...routeManifest[key],
      } as NestedRouteManifest;
      const children = formatNestedRouteManifest(routeManifest, route.id);
      if (children.length > 0) route.children = children;
      return route;
    });
}

function defineConventionalRoutes(
  rootDir: string,
  ignoredFilePatterns?: string[],
): RouteManifest {
  const files: { [routeId: string]: string } = {};

  // 1. find all route components in src/pages
  visitFiles(
    path.join(rootDir, 'src', 'pages'),
    file => {
      if (
        ignoredFilePatterns &&
        ignoredFilePatterns.some(pattern => minimatch(file, pattern))
      ) {
        return;
      }

      const filePath = path.join('pages', file);
      if (isRouteModuleFile(file)) {
        let routeId = createRouteId(filePath);
        files[routeId] = filePath;
        return;
      }

      throw new Error(
        `Invalid route module file: ${path.join(rootDir, filePath)}`,
      );
    },
  );

  const routeIds = Object.keys(files).sort(byLongestFirst);

  const uniqueRoutes = new Map<string, string>();

  // 2. recurse through all routes using the public defineRoutes() API
  function defineNestedRoutes(
    defineRoute: DefineRouteFunction,
    parentId?: string,
  ): void {
    const childRouteIds = routeIds.filter((id) => {
      const parentRouteId = findParentRouteId(routeIds, id);
      return parentRouteId === parentId;
    });

    console.log('childRouteIds: ', childRouteIds);

    for (let routeId of childRouteIds) {
      const routePath: string | undefined = createRoutePath(
        routeId.slice((removeLayoutStrFromId(parentId) || 'pages').length),
      );

      const isIndexRoute = routeId.endsWith('/index');
      let fullPath = createRoutePath(routeId.slice('pages'.length + 1));
      let uniqueRouteId = (fullPath || '') + (isIndexRoute ? '?index' : '');

      if (uniqueRouteId) {
        if (uniqueRoutes.has(uniqueRouteId)) {
          throw new Error(
            `Path ${JSON.stringify(fullPath)} defined by route ${JSON.stringify(
              routeId,
            )} conflicts with route ${JSON.stringify(
              uniqueRoutes.get(uniqueRouteId),
            )}`,
          );
        } else {
          uniqueRoutes.set(uniqueRouteId, routeId);
        }
      }

      if (isIndexRoute) {
        let invalidChildRoutes = routeIds.filter(
          (id) => findParentRouteId(routeIds, id) === routeId,
        );

        if (invalidChildRoutes.length > 0) {
          throw new Error(
            `Child routes are not allowed in index routes. Please remove child routes of ${routeId}`,
          );
        }

        defineRoute(routePath, files[routeId], {
          index: true,
        });
      } else {
        defineRoute(routePath, files[routeId], () => {
          defineNestedRoutes(defineRoute, routeId);
        });
      }
    }
  }

  return defineRoutes(defineNestedRoutes);
}

let escapeStart = '[';
let escapeEnd = ']';

export function createRoutePath(partialRouteId: string): string | undefined {
  let result = '';
  let rawSegmentBuffer = '';

  let inEscapeSequence = 0;
  let skipSegment = false;

  partialRouteId = removeLayoutStrFromId(partialRouteId);

  for (let i = 0; i < partialRouteId.length; i++) {
    const char = partialRouteId.charAt(i);
    const lastChar = i > 0 ? partialRouteId.charAt(i - 1) : undefined;
    const nextChar = i < partialRouteId.length - 1 ? partialRouteId.charAt(i + 1) : undefined;

    function isNewEscapeSequence() {
      return !inEscapeSequence && char === escapeStart && lastChar !== escapeStart;
    }

    function isCloseEscapeSequence() {
      return inEscapeSequence && char === escapeEnd && nextChar !== escapeEnd;
    }

    // src/pages/__app/dashboard.tsx || src/pages/__app.tsx
    function isStartOfLayoutSegment() {
      return char === '_' && nextChar === '_' && !rawSegmentBuffer;
    }

    if (skipSegment) {
      if (char === '/' || char === '.' || char === path.win32.sep) {
        skipSegment = false;
      }
      continue;
    }

    if (isNewEscapeSequence()) {
      inEscapeSequence++;
      continue;
    }

    if (isCloseEscapeSequence()) {
      inEscapeSequence--;
      continue;
    }

    // if you want to mark resource route for a `/sitemap.xml`, add a route `src/pages/[sitemap.xml].tsx`
    if (inEscapeSequence) {
      result += char;
      continue;
    }

    if (char === '/' || char === path.win32.sep || char === '.') {
      if (rawSegmentBuffer === 'index' && result.endsWith('index')) {
        result = result.replace(/\/?index$/, '');
      } else {
        result += '/';
      }
      rawSegmentBuffer = '';
      continue;
    }

    if (isStartOfLayoutSegment()) {
      skipSegment = true;
      continue;
    }

    rawSegmentBuffer += char;

    if (char === '$') {
      result += typeof nextChar === 'undefined' ? '*' : ':';
      continue;
    }

    result += char;
  }

  if (rawSegmentBuffer === 'index' && result.endsWith('index')) {
    result = result.replace(/\/?index$/, '');
  }

  return result || undefined;
}

function findParentRouteId(
  routeIds: string[],
  childRouteId: string,
): string | undefined {
  // return routeIds.find((id) => childRouteId.startsWith(`${id}/`));

  return routeIds.find((id) => {
    return (
      // childRouteId.startsWith(`${id}/`)
      childRouteId !== id && id.endsWith('layout') && childRouteId.startsWith(`${id.slice(0, id.length - '/layout'.length)}`)
    );
  });
}

function byLongestFirst(a: string, b: string): number {
  return b.length - a.length;
}

function visitFiles(
  dir: string,
  visitor: (file: string) => void,
  baseDir = dir,
): void {
  for (let filename of fs.readdirSync(dir)) {
    let file = path.resolve(dir, filename);
    let stat = fs.lstatSync(file);

    if (stat.isDirectory()) {
      visitFiles(file, visitor, baseDir);
    } else if (stat.isFile()) {
      visitor(path.relative(baseDir, file));
    }
  }
}

const entryExts = ['.js', '.jsx', '.ts', '.tsx'];

function findGlobalLayout(srcDir: string, basename: string): string | undefined {
  for (let ext of entryExts) {
    let file = path.resolve(srcDir, basename + ext);
    if (fs.existsSync(file)) return path.relative(srcDir, file);
  }

  return undefined;
}

/**
 * remove `/layout` str if the routeId has it
 *
 * /About/layout -> /About
 * /About/layout/index -> /About/layout/index
 */
function removeLayoutStrFromId(id?: string) {
  if (!id) {
    return id;
  }
  return id.endsWith('/layout') ? id.slice(0, id.length - '/layout'.length) : id;
}