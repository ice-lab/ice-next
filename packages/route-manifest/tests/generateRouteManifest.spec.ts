import { expect, test, describe } from 'vitest';
import path from 'path';
import { fileURLToPath } from 'url';
import { generateRouteManifest } from '../src/index';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const fixturesDir = path.join(__dirname, 'fixtures');

describe('generateRouteManifest function', () => {
  test('basic-routes', () => {
    const routeManifest = generateRouteManifest(path.join(fixturesDir, 'basic-routes'));
    expect(routeManifest).toMatchSnapshot();
  });

  test('dynamic-routes', () => {
    const routeManifest = generateRouteManifest(path.join(fixturesDir, 'dynamic-routes'));
    expect(routeManifest).toMatchSnapshot();
  });

  test('layout-routes', () => {
    const routeManifest = generateRouteManifest(path.join(fixturesDir, 'layout-routes'));
    expect(routeManifest).toMatchSnapshot();
  });

  test('doc-delimeters-routes', () => {
    const routeManifest = generateRouteManifest(path.join(fixturesDir, 'dot-delimeters-routes'));
    expect(routeManifest).toMatchSnapshot();
  });

  test('splat-routes', () => {
    const routeManifest = generateRouteManifest(path.join(fixturesDir, 'splat-routes'));
    expect(routeManifest).toMatchSnapshot();
  });

  test('invalid-routes', () => {
    expect(() => generateRouteManifest(path.join(fixturesDir, 'invalid-routes'))).toThrow(`invalid character in 'pages/[a.pdf]'. Only support char: -, \\w, /`);
  });
});
