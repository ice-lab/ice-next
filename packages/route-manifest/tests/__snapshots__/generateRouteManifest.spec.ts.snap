// Vitest Snapshot v1

exports[`generateRouteManifest function > basic-routes 1`] = `
{
  "pages/About/index": {
    "componentName": "PagesAboutIndex",
    "file": "pages/About/index.tsx",
    "id": "pages/About/index",
    "index": true,
    "parentId": "pages/layout",
    "path": "/About",
  },
  "pages/About/me/index": {
    "componentName": "PagesAboutMeIndex",
    "file": "pages/About/me/index.tsx",
    "id": "pages/About/me/index",
    "index": true,
    "parentId": "pages/layout",
    "path": "/About/me",
  },
  "pages/home": {
    "componentName": "PagesHome",
    "file": "pages/home.tsx",
    "id": "pages/home",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/home",
  },
  "pages/index": {
    "componentName": "PagesIndex",
    "file": "pages/index.tsx",
    "id": "pages/index",
    "index": true,
    "parentId": "pages/layout",
    "path": undefined,
  },
  "pages/layout": {
    "componentName": "PagesLayout",
    "file": "pages/layout.tsx",
    "id": "pages/layout",
    "index": undefined,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function > doc-delimeters-routes 1`] = `
{
  "pages/home.news": {
    "componentName": "PagesHomeNews",
    "file": "pages/home.news.tsx",
    "id": "pages/home.news",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/home/news",
  },
  "pages/layout": {
    "componentName": "PagesLayout",
    "file": "pages/layout.tsx",
    "id": "pages/layout",
    "index": undefined,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function > dynamic-routes 1`] = `
{
  "pages/about": {
    "componentName": "PagesAbout",
    "file": "pages/about.tsx",
    "id": "pages/about",
    "index": undefined,
    "parentId": undefined,
    "path": "/about",
  },
  "pages/blog/$id": {
    "componentName": "PagesBlog$id",
    "file": "pages/blog/$id.tsx",
    "id": "pages/blog/$id",
    "index": undefined,
    "parentId": undefined,
    "path": "/blog/:id",
  },
  "pages/blog/index": {
    "componentName": "PagesBlogIndex",
    "file": "pages/blog/index.tsx",
    "id": "pages/blog/index",
    "index": true,
    "parentId": undefined,
    "path": "/blog",
  },
  "pages/index": {
    "componentName": "PagesIndex",
    "file": "pages/index.tsx",
    "id": "pages/index",
    "index": true,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function > layout-routes 1`] = `
{
  "pages/about": {
    "componentName": "PagesAbout",
    "file": "pages/about.tsx",
    "id": "pages/about",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/about",
  },
  "pages/blog/$id": {
    "componentName": "PagesBlog$id",
    "file": "pages/blog/$id.tsx",
    "id": "pages/blog/$id",
    "index": undefined,
    "parentId": "pages/blog/layout",
    "path": "/:id",
  },
  "pages/blog/index": {
    "componentName": "PagesBlogIndex",
    "file": "pages/blog/index.tsx",
    "id": "pages/blog/index",
    "index": true,
    "parentId": "pages/blog/layout",
    "path": undefined,
  },
  "pages/blog/layout": {
    "componentName": "PagesBlogLayout",
    "file": "pages/blog/layout.tsx",
    "id": "pages/blog/layout",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/blog",
  },
  "pages/home/index": {
    "componentName": "PagesHomeIndex",
    "file": "pages/home/index.tsx",
    "id": "pages/home/index",
    "index": true,
    "parentId": "pages/home/layout",
    "path": undefined,
  },
  "pages/home/layout": {
    "componentName": "PagesHomeLayout",
    "file": "pages/home/layout.tsx",
    "id": "pages/home/layout",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/home",
  },
  "pages/home/layout/index": {
    "componentName": "PagesHomeLayoutIndex",
    "file": "pages/home/layout/index.tsx",
    "id": "pages/home/layout/index",
    "index": true,
    "parentId": "pages/home/layout",
    "path": "/layout",
  },
  "pages/index": {
    "componentName": "PagesIndex",
    "file": "pages/index.tsx",
    "id": "pages/index",
    "index": true,
    "parentId": "pages/layout",
    "path": undefined,
  },
  "pages/layout": {
    "componentName": "PagesLayout",
    "file": "pages/layout.tsx",
    "id": "pages/layout",
    "index": undefined,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function > splat-routes 1`] = `
{
  "pages/$": {
    "componentName": "Pages$",
    "file": "pages/$.tsx",
    "id": "pages/$",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/*",
  },
  "pages/home": {
    "componentName": "PagesHome",
    "file": "pages/home.tsx",
    "id": "pages/home",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/home",
  },
  "pages/layout": {
    "componentName": "PagesLayout",
    "file": "pages/layout.tsx",
    "id": "pages/layout",
    "index": undefined,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function basic-routes 1`] = `
Object {
  "pages/About/index": Object {
    "componentName": "PagesAboutIndex",
    "file": "pages/About/index.tsx",
    "id": "pages/About/index",
    "index": true,
    "parentId": "pages/layout",
    "path": "/About",
  },
  "pages/About/me/index": Object {
    "componentName": "PagesAboutMeIndex",
    "file": "pages/About/me/index.tsx",
    "id": "pages/About/me/index",
    "index": true,
    "parentId": "pages/layout",
    "path": "/About/me",
  },
  "pages/home": Object {
    "componentName": "PagesHome",
    "file": "pages/home.tsx",
    "id": "pages/home",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/home",
  },
  "pages/index": Object {
    "componentName": "PagesIndex",
    "file": "pages/index.tsx",
    "id": "pages/index",
    "index": true,
    "parentId": "pages/layout",
    "path": undefined,
  },
  "pages/layout": Object {
    "componentName": "PagesLayout",
    "file": "pages/layout.tsx",
    "id": "pages/layout",
    "index": undefined,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function doc-delimeters-routes 1`] = `
Object {
  "pages/home.news": Object {
    "componentName": "PagesHomeNews",
    "file": "pages/home.news.tsx",
    "id": "pages/home.news",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/home/news",
  },
  "pages/layout": Object {
    "componentName": "PagesLayout",
    "file": "pages/layout.tsx",
    "id": "pages/layout",
    "index": undefined,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function dynamic-routes 1`] = `
Object {
  "pages/about": Object {
    "componentName": "PagesAbout",
    "file": "pages/about.tsx",
    "id": "pages/about",
    "index": undefined,
    "parentId": undefined,
    "path": "/about",
  },
  "pages/blog/$id": Object {
    "componentName": "PagesBlog$id",
    "file": "pages/blog/$id.tsx",
    "id": "pages/blog/$id",
    "index": undefined,
    "parentId": undefined,
    "path": "/blog/:id",
  },
  "pages/blog/index": Object {
    "componentName": "PagesBlogIndex",
    "file": "pages/blog/index.tsx",
    "id": "pages/blog/index",
    "index": true,
    "parentId": undefined,
    "path": "/blog",
  },
  "pages/index": Object {
    "componentName": "PagesIndex",
    "file": "pages/index.tsx",
    "id": "pages/index",
    "index": true,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function layout-routes 1`] = `
Object {
  "pages/about": Object {
    "componentName": "PagesAbout",
    "file": "pages/about.tsx",
    "id": "pages/about",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/about",
  },
  "pages/blog/$id": Object {
    "componentName": "PagesBlog$id",
    "file": "pages/blog/$id.tsx",
    "id": "pages/blog/$id",
    "index": undefined,
    "parentId": "pages/blog/layout",
    "path": "/:id",
  },
  "pages/blog/index": Object {
    "componentName": "PagesBlogIndex",
    "file": "pages/blog/index.tsx",
    "id": "pages/blog/index",
    "index": true,
    "parentId": "pages/blog/layout",
    "path": undefined,
  },
  "pages/blog/layout": Object {
    "componentName": "PagesBlogLayout",
    "file": "pages/blog/layout.tsx",
    "id": "pages/blog/layout",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/blog",
  },
  "pages/home/index": Object {
    "componentName": "PagesHomeIndex",
    "file": "pages/home/index.tsx",
    "id": "pages/home/index",
    "index": true,
    "parentId": "pages/home/layout",
    "path": undefined,
  },
  "pages/home/layout": Object {
    "componentName": "PagesHomeLayout",
    "file": "pages/home/layout.tsx",
    "id": "pages/home/layout",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/home",
  },
  "pages/home/layout/index": Object {
    "componentName": "PagesHomeLayoutIndex",
    "file": "pages/home/layout/index.tsx",
    "id": "pages/home/layout/index",
    "index": true,
    "parentId": "pages/home/layout",
    "path": "/layout",
  },
  "pages/index": Object {
    "componentName": "PagesIndex",
    "file": "pages/index.tsx",
    "id": "pages/index",
    "index": true,
    "parentId": "pages/layout",
    "path": undefined,
  },
  "pages/layout": Object {
    "componentName": "PagesLayout",
    "file": "pages/layout.tsx",
    "id": "pages/layout",
    "index": undefined,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function splat-routes 1`] = `
Object {
  "pages/$": Object {
    "componentName": "Pages$",
    "file": "pages/$.tsx",
    "id": "pages/$",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/*",
  },
  "pages/home": Object {
    "componentName": "PagesHome",
    "file": "pages/home.tsx",
    "id": "pages/home",
    "index": undefined,
    "parentId": "pages/layout",
    "path": "/home",
  },
  "pages/layout": Object {
    "componentName": "PagesLayout",
    "file": "pages/layout.tsx",
    "id": "pages/layout",
    "index": undefined,
    "parentId": undefined,
    "path": undefined,
  },
}
`;
